DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c

```diff
@@ -861,29 +865,392 @@ QueryVariableInfoHook (
+
+  // 1. Locate Protocols
+  Status = gBS->LocateProtocol(&gEfiGraphicsOutputProtocolGuid, NULL, (VOID **)&Gop);
+  if (EFI_ERROR(Status)) return;
+
+  Status = gBS->LocateProtocol(&gEfiHiiFontProtocolGuid, NULL, (VOID **)&HiiFont);
+  if (EFI_ERROR(Status)) return;
+
+  ScreenWidth  = Gop->Mode->Info->HorizontalResolution;
+  ScreenHeight = Gop->Mode->Info->VerticalResolution;
+
+  // 2. Calculate Dimensions
+  // If RequestedWidth is 0, we must measure the strings.
+  // If provided, RequestedWidth is usually in "Columns" (chars).
+
+  UINTN MaxStrLen = 0;
+
+  VA_START(Args, NumberOfLines);
+  VA_COPY(ArgsCopy, Args);
+  for (Index = 0; Index < NumberOfLines; Index++) {
+    String = VA_ARG(Args, CHAR16 *);
+    if (String != NULL) {
+      UINTN Len = StrLen(String);
+      if (Len > MaxStrLen) MaxStrLen = Len;
+    }
+  }
+  VA_END(Args);
+
+  if (RequestedWidth == 0) {
+    RequestedWidth = MaxStrLen;
+  }
+
+  // Enforce a minimum width if needed, or cap to screen width logic
+  if (RequestedWidth < MaxStrLen) RequestedWidth = MaxStrLen;
+
+  // Calculate pixel dimensions based on 8x19 standard glyphs
+  UnscaledContentW = RequestedWidth * GLYPH_WIDTH;
+  UnscaledContentH = NumberOfLines * GLYPH_HEIGHT;
+
+  // Final Box Size (Scaled Content + Scaled Padding)
+  // We apply scale factor to everything to ensure crisp look
+  BoxW = (UnscaledContentW * ScaleFactor) + (PADDING_X * ScaleFactor * 2);
+  BoxH = (UnscaledContentH * ScaleFactor) + (PADDING_Y * ScaleFactor * 2);
+
+  // Center on screen
+  BoxX = (ScreenWidth > BoxW) ? (ScreenWidth - BoxW) / 2 : 0;
+  BoxY = (ScreenHeight > BoxH) ? (ScreenHeight - BoxH) / 2 : 0;
+
+  // 3. Save Background
+  BackBuffer = AllocateZeroPool(BoxW * BoxH * sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
+  if (BackBuffer != NULL) {
+    Gop->Blt(Gop, BackBuffer, EfiBltVideoToBltBuffer, BoxX, BoxY, 0, 0, BoxW, BoxH, 0);
+  }
+
+  // 4. Draw Box & Border
+  EFI_GRAPHICS_OUTPUT_BLT_PIXEL BgColor = POPUP_BG_COLOR;
+  EFI_GRAPHICS_OUTPUT_BLT_PIXEL BorderColor = POPUP_BORDER_COLOR;
+
+  // Solid Fill
+  Gop->Blt(Gop, &BgColor, EfiBltVideoFill, 0, 0, BoxX, BoxY, BoxW, BoxH, 0);
+
+  // Borders
+  Gop->Blt(Gop, &BorderColor, EfiBltVideoFill, 0, 0, BoxX, BoxY, BoxW, BORDER_WIDTH, 0); // Top
+  Gop->Blt(Gop, &BorderColor, EfiBltVideoFill, 0, 0, BoxX, BoxY + BoxH - BORDER_WIDTH, BoxW, BORDER_WIDTH, 0); // Bottom
+  Gop->Blt(Gop, &BorderColor, EfiBltVideoFill, 0, 0, BoxX, BoxY, BORDER_WIDTH, BoxH, 0); // Left
+  Gop->Blt(Gop, &BorderColor, EfiBltVideoFill, 0, 0, BoxX + BoxW - BORDER_WIDTH, BoxY, BORDER_WIDTH, BoxH, 0); // Right
+
+  // 5. Prepare for Text Rendering
+  // We allocate enough space for one line at a time
+  UINTN LinePixels1x = UnscaledContentW * GLYPH_HEIGHT;
+  RenderBuffer1x = AllocateZeroPool(LinePixels1x * sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
+  RenderBufferScaled = AllocateZeroPool(LinePixels1x * ScaleFactor * ScaleFactor * sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));

```

Comment:
Something seems to be trashing memory:
```
DXE_ASSERT!: [DxeCore] .../coreboot/payloads/external/edk2/workspace/Dasharo/MdeModulePkg/Core/Dxe/Mem/Pool.c (741): Tail->Signature == ((('p') | ('t' << 8)) | ((('a') | ('l' << 8)) << 16))

DXE_ASSERT!: [DxeCore] .../coreboot/payloads/external/edk2/workspace/Dasharo/MdeModulePkg/Core/Dxe/Mem/Pool.c (742): Head->Size == Tail->Size

ASSERT_EFI_ERROR (Status = Invalid Parameter)

DXE_ASSERT!: [FmpDxe] .../coreboot/payloads/external/edk2/workspace/Dasharo/MdePkg/Library/UefiMemoryAllocationLib/MemoryAllocationLib.c (821): !(((RETURN_STATUS)(Status)) >= 0x8000000000000000ULL)
```

If I allocate 10 times the amount on this line the errors are gone, but I don't see this buffer being overrun in the code.  Maybe the issue in a different place.